### PR TITLE
Fix/27 concurrent modification

### LINKED ISSUE #27: un/subscribe during `newState` callback causes crash -> ConcurrentModificationException
Hi,

I'm toying with ReKotlin and ran into a problem:
I have an Android app with a single Activity, that activity displays different screens by attaching and detaching views from its own view (just a container).

To navigate between different screens I dispatch actions → trigger a state change (the current route changes) → the activity detaches old screen & attaches new screen
Each of these screens subscribes to the store.

So I get the following sequence:

```
activity        store
    | subscribe() |
    |------------>|
    |             |  dispatch()
    |             |<----------------
    | newState()  |
   ||<------------|  
   ||             |   create
   ||-------------|--------------> screen
   ||             | subscribe()      |
   ||             |<-----------------|
   ||<------------|
    |             |
    |             |
    💥            💥
```

this results in a `ConcurrentModificationException`, full stacktrace:

```
java.util.ConcurrentModificationException
        at java.util.ArrayList$Itr.next(ArrayList.java:831)
        at org.rekotlin.Store.set_state(Store.kt:163)
        at org.rekotlin.Store._defaultDispatch(Store.kt:129)
        at org.rekotlin.Store$dispatchFunction$1.invoke(Store.kt:51)
        at org.rekotlin.Store$dispatchFunction$1.invoke(Store.kt:27)
        at my.redux.playground.MainActivityKt$loggingMiddleware$1$1$1.invoke(MainActivity.kt:303)
        at my.redux.playground.MainActivityKt$loggingMiddleware$1$1$1.invoke(MainActivity.kt)
        at org.rekotlin.Store.dispatch(Store.kt:133)
        at my.redux.playground.ScreenOne$4.invoke(MainActivity.kt:108)
        at my.redux.playground.ScreenOne$4.invoke(MainActivity.kt:92)
        at my.redux.playground.MainActivityKt$doOnClick$1.onClick(MainActivity.kt:329)
        at android.view.View.performClick(View.java:5637)
        at android.view.View$PerformClick.run(View.java:22429)
        at android.os.Handler.handleCallback(Handler.java:751)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6119)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

The problem seems to be caused by calling `Store#subscribe` or `Store#unsubscribe` during execution of state updates (in [Store.kt lines 39-41](https://github.com/ReKotlin/ReKotlin/blob/master/src/main/kotlin/org/rekotlin/Store.kt#L39-L41) ).
The list of subscriptions` is mutated while iterated over.

My assumption was that subscribing and unsubscribing is independent of state updates and expected that a subscription would only get "active" after/inbetween state changes. Is my assumption wrong? Or is this a bug? Am i using ReKotlin in an unexpected way - if so is there a better approach?

Thanks guys 🙏

### PR DESCRIPTION
**Pull request by NemoOudeis**
As described in #27 any call to `subscribe` or `unsubscribe` from a `newState` would create a `ConcurrentModificationException` because the list of `subscriptions` was modified (by un/subscribe) while iterating over it (from the setter of `_state`).

By using a `CopyOnWriteArrayList` the iteration over the original list finishes (without modification) while any future access of `subscription` reflects the mutation.

Fixes #27

### CODE REVIEW

**Comment by kaustavjaiswal** [comment]
LGTM

**Comment by jitinsharma** [comment]
I have doubts over performance because [CopyOnWriteArrayList](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html) would create a copy everytime subscribe/unsubscribe is called of a list which carries all active subscriptions. Also since by definition it is thread safe, it would be slower than general arraylist.

**Comment by kaustavjaiswal** [comment]
While the performance issue is a concern in my mind too, the only other way out seems like a synchronized list which has a worse performance characteristic in case of more read operations, which would be the case I'm guessing, with the CopyOverWriteArrayList(COWAL) the read operations will remain more performant but the writes will suffer.

Should be fine for lower number of concurrent writes, but will slow down significantly for larger number because of the way JVM handles the COWAL

**Comment by NemoOudeis** [comment]
I agree that synchronization incurs performance overhead. 
If we don't rely on COWAL we have to put in explicit locking whenever we access `subscriptions` - which also has performance overhead **and** makes the code harder to understand & reason about. The performance overhead might be less though, because we don't need to copy if we properly guard against concurrent access.

**Comment by kaustavjaiswal** [comment]
@kmmraj Waiting for your inputs on this, to confirm, merge and close the issue.

File: src/test/kotlin/org/rekotlin/TestFakes.kt [hunk 1]
```diff
@@ -114,6 +114,38 @@ internal class TestStoreSubscriber<T> : StoreSubscriber<T> {
...
 
+/**
+ * A subscriber contains another sub-subscribers [StoreSubscriber], which could be subscribe/unsubscribe at some point
+ */
+internal class ViewSubscriberTypeA(var store: Store<TestStringAppState>) : StoreSubscriber<TestStringAppState> {
+    private val viewB by lazy { ViewSubscriberTypeB(store) }
+    private val viewC by lazy { ViewSubscriberTypeC() }
+
+    override fun newState(state: TestStringAppState) {
+        when(state.testValue){
```
Comment (kmmraj) [inline src/test/kotlin/org/rekotlin/TestFakes.kt]: I am not sure why to subscribe or unsubscribe in a newState function.
A class should subscribe to the state when it is getting created and unsubscribe when it is getting destroyed.
Any conditional subscription or unsubscribing should be delegated to the separate class by a dispatch action that triggers that class creation.

@jitinsharma, @kaustavjaiswal -- I need your view before merging this.

Anyway, in meanwhile @NemoOudeis Could you please add to the README.md, that number of concurrent writes should be less than the number of concurrent reads in the subscriptions list.

**Comment by kaustavjaiswal** [comment]
@NemoOudeis The README.md update to mention the existence of COWAL and comment above the COWAL declaration to ensure concurrent reads should be always greater than writes should be good enough to warn library users and avoid OOM.

File: src/test/kotlin/org/rekotlin/TestFakes.kt (hunk 1 above)
Comment (jitinsharma) [reply src/test/kotlin/org/rekotlin/TestFakes.kt]: I have also doubts on scenarios where a subscription is required on new state function.

Comment (kaustavjaiswal) [reply src/test/kotlin/org/rekotlin/TestFakes.kt]: @NemoOudeis Would it be possible for you to mention an example of real-world usage (if possible) within your application where such behavior is required?

**Comment by NemoOudeis** [comment]
I will update the `README.md` to mention the issue.

File: src/test/kotlin/org/rekotlin/TestFakes.kt (hunk 1 above)
Comment (NemoOudeis) [reply src/test/kotlin/org/rekotlin/TestFakes.kt]: I can describe an example, the code is closed source, so I cannot share it here.

We use a component (think react components, i.e. UI with some behavior) hierarchy that assembles the UI for the user, an example of such a hierarchy could be:
```
     +------- Screen Component -------+
     |               |                |
     |               |                |
Component A          |            Component B
                     |
                Compontent C
```

Each of the components subscribes to (sub-)states and behaves accordingly.
In cases where the hierarchy is known at compile time it's easy, we can just create and the components A, B and C and subscribe them at our own convenience.

The problem occurs when we don't know the hierarchy at compile time, e.g. a components should be shown (or not) based on some data fetched from a server. In that case the data is fetched -> send to the store -> reducers update state -> store notifies subscribers. In that notification (`newState) the `Screen Component` decides to attach (or detach) another component - which requires that component to subscribe unsubscribe (this happens in the call stack of `newState`).

Does that make sense?

File: README.md [hunk 2]
```diff
@@ -247,6 +247,12 @@ val store = Store(
 	automaticallySkipRepeats = false)
 ```
 
+### Subscribe/Unsubscribe during `newState`
+
+Under the hood ReKotlin uses a [`CopyOnWriteArrayList`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html) to manage subscriptions (see [PR 29](https://github.com/ReKotlin/ReKotlin/pull/29) for more details). This implementation detail means that then number of concurrent writes to the subscriptions should be less than the number of concurrent reads.
```
Comment (kaustavjaiswal) [inline README.md]: *the number of

**Comment by kaustavjaiswal** [comment]
@NemoOudeis Can you take a fresh pull, to ensure the build is successful, some issues on Travis Side for Oracle JDK, I have fixed them in my latest merged PR

File: README.md (hunk 2 above)
Comment (NemoOudeis) [reply README.md]: fixed

**Pull request closed without merging**